cluster-autoscaler/core/utils/utils.go

```diff
@@ -42,7 +42,7 @@ import (
 func GetNodeInfosForGroups(nodes []*apiv1.Node, nodeInfoCache map[string]*schedulerframework.NodeInfo, cloudProvider cloudprovider.CloudProvider, listers kube_util.ListerRegistry,
 	// TODO(mwielgus): This returns map keyed by url, while most code (including scheduler) uses node.Name for a key.
 	// TODO(mwielgus): Review error policy - sometimes we may continue with partial errors.
-	daemonsets []*appsv1.DaemonSet, predicateChecker simulator.PredicateChecker, ignoredTaints taints.TaintKeySet) (map[string]*schedulerframework.NodeInfo, errors.AutoscalerError) {
+	daemonsets []*appsv1.DaemonSet, predicateChecker simulator.PredicateChecker, ignoredTaints taints.TaintKeySet, useTemplatesOnly bool) (map[string]*schedulerframework.NodeInfo, errors.AutoscalerError) {
 	result := make(map[string]*schedulerframework.NodeInfo)
```

Comment:
There is lots of half done interactions with cache/result when useTemplatesOnly is used. Like populating cache and not using it. Using "last resort" nodes for if no teplate is provided. It would be cleaner to have a IF userTemplateOnly { } and the very beginning of the function, try generate stuff based on templates only and return.

Reply:
Right that suggestion makes it more clean and clear, while avoiding processing things we don't use (I tried to minimize the needed changes but that's not a great criterion indeed).

Updated to implement it as suggested. While doing so, had to make those two choices that could benefit from second opinion:
* Not using nodeInfoCache. We weren't using it for templates before either, but I suppose it was likely omitted to allow adoption of a real-world node as a model whenever possible (so might not be optimal here). On Azure and GCE GetNodeInfoFromTemplate->TemplateNodeInfo hits caches but didn't dig for other providers. Wdyt?
* Left cloud-providers the option to selectively return ErrNotImplemented depending on the group (otherwise error handling could be simpler)